Feature/frontend create and delete relationship

manwithtwowatches opened the pull request:
https://www.openproject.org/work_packages/13948
https://www.openproject.org/work_packages/13949

Commit:
Added endpoint for creating relations

Commit:
Added endpoint for deleting relationships

Commit:
Added authorization to creating and deleting relations

Commit:
Add create and delete relation links to representers.

Commit:
Use old prototype autocomplete in relations tab.

manwithtwowatches commented on app/assets/javascripts/angular/work_packages/tabs/work-package-relation-directive.js:
```diff
@@ -53,6 +62,23 @@ angular.module('openproject.uiComponents')
         setExpandState();
       });
 
+      // Massive hack alert - Using old prototype autocomplete ///////////
```
@myabc @ulferts Opinions wanted - Is this (or something like this) an acceptable temporary solution for this sprint? For me it doesn't really make sense reimplementing it until we have an api endpoint for work packages index.

myabc replied:
@manwithtwowatches IMO, yes

Commit:
WIP Refactoring relations directives to correspond to relations not work packages.

This seems like an awfully long way round now to achieve having the activity object and thus its delete link available in the directive. Have also broken parent and child relations since they are not actually relations.

Commit:
WIP sketchily wired up create/delete relations. No tests or validation handling.

Commit:
Add separate directive for child work packages even though it looks awfully similar to relations.

Commit:
Wire back up read only children/parent and add child button linking to old create view.

Commit:
Some basic validation error handling in angular for api requests.

Commit:
WIP trying to fix all the karma tests. not going too well.

Commit:
Having a break from trying to test relations directive as i'm demoralised.

Commit:
Using a $timeout approach to test relations directive.

Commit:
Add permissions for add/remove relations.

Commit:
Did sneaky css tricks to get the relations autocomplete looking okish.

Commit:
Merge branch 'dev' into feature/frontend-create-and-delete-relationship

Conflicts:
	lib/api/v3/work_packages/work_package_representer.rb
	spec/lib/api/v3/work_packages/work_package_representer_spec.rb

Commit:
Put back in missing dev after merge cockup.

hschink commented on lib/services/create_relation.rb:
```diff
@@ -0,0 +1,18 @@
+    if @relation.save
+      success.(created: true)
+    else
+      binding.pry
```
eeek

hschink commented on public/templates/work_packages/tabs/relations.html:
```diff
@@ -1,58 +1,75 @@
+                           work-package="workPackage"
+                           relations="relatedTo"
+                           relation-identifier="'relates'"
+                           button-title="Add Related to"
```
Indentation... see the other ones too.

hschink commented on public/templates/work_packages/tabs/relations.html:
```diff
@@ -1,58 +1,75 @@
+                           relation-identifier="'parent'"
                            button-title="Change Parent"
                            button-icon="<i class='icon-hierarchy icon-edit'></i>"
                            singleton-relation="true">
```
Indentation

hschink commented on app/assets/javascripts/angular/work_packages/tabs/related-work-package-table-row-directive.js:
```diff
@@ -27,45 +27,36 @@
+        WorkPackageService.removeWorkPackageRelation(scope.relation).then(function(response){
+          scope.$emit('workPackageRefreshRequired', '');
+        }, function(error) {
+          ApiHelper.handleError(scope, error);
```
This is not defined.

hschink commented on lib/api/v3/relations/relations_api.rb:
```diff
@@ -0,0 +1,45 @@
+            end
+          end
+
+          namespace ':id' do
```
That won't work in the context of another API that already introduces `:id` namespace (e.g., work packages). Thus, we should rename this parameter to `:relation_id` or something similar.

Commit:
Clean up the pr a bit.

Commit:
Track down super irritating related to/by bug

Commit:
Rename relations api namespace from id to relation_id.

hschink commented:
Ready for merge as soon as the tests pass! :+1:
